# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaManager.cpp
# Commit: 9b14e7bdee7a
# Full Hash: 9b14e7bdee7a21cbd50f1cb0d57faa88a7025061
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-10-17 21:54:25
# Regressor Bug: 1397528
# File Overlap Count: 1
# Description:
#   Bug 1397528 - Handle cleaned up windows in async enumerateDevices response. r=jib
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D48661
# ==============================================================================

diff -r 58b4283727ea -r 9b14e7bdee7a dom/media/MediaManager.cpp
--- a/dom/media/MediaManager.cpp	Wed Oct 16 19:08:46 2019 +0000
+++ b/dom/media/MediaManager.cpp	Thu Oct 17 07:57:13 2019 +0000
@@ -3248,7 +3248,14 @@
                               devices)
       ->Then(
           GetCurrentThreadSerialEventTarget(), __func__,
-          [windowListener, sourceListener, devices](bool) {
+          [self = RefPtr<MediaManager>(this), this, windowListener,
+           sourceListener, devices](bool) {
+            if (!IsWindowListenerStillActive(windowListener)) {
+              MOZ_ASSERT(!windowListener->Remove(sourceListener));
+              return DevicesPromise::CreateAndReject(
+                  MakeRefPtr<MediaMgrError>(MediaMgrError::Name::AbortError),
+                  __func__);
+            }
             DebugOnly<bool> rv = windowListener->Remove(sourceListener);
             MOZ_ASSERT(rv);
             return DevicesPromise::CreateAndResolve(devices, __func__);
@@ -3256,9 +3263,7 @@
           [windowListener, sourceListener](RefPtr<MediaMgrError>&& aError) {
             // This may fail, if a new doc has been set the OnNavigation
             // method should have removed all previous active listeners.
-            // Attempt to clean it here, just in case, but ignore the return
-            // value.
-            Unused << windowListener->Remove(sourceListener);
+            MOZ_ASSERT(!windowListener->Remove(sourceListener));
             return DevicesPromise::CreateAndReject(std::move(aError), __func__);
           });
 }
